Fixed #35515 -- Added auto-importing to shell command. (#18158)
django/django: docs/howto/overriding-shell.txt

=========================

> How to override the shell

Reviewer: Let’s retitle the document “How to customize the `shell` command”. Whilst we do **override** the management command when making customizations, it’s not the clearest term.

Also, add a link from the `shell` section in `ref/django-admin.txt` to this document, with some sentence about how it’s possible to customize the shell.

=========================

> Django provides functionality to auto-import project models in the shell. In your project, you might want to either avoid importing models or include additional imports in the shell. You can do this by subclassing the :djadmin:`shell` command and overriding a method.

Reviewer: Wrap lines at 80 characters: https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#guidelines-for-restructuredtext-files

Most text editors allow configuring a line to appear in that column, so you can visually wrap it. You can also find automatic wrapping commands or extensions for most editors.

> Setting up the app directory

Reviewer: I think we can drop this section and just have a one-sentence link to `howto/custom-management-commands`. Developers doing shell customization presumably have some level of experience and don’t need everything spelled out.

Setting up the App Directory
============================

Following the :doc:`guide </howto/custom-management-commands>` on how to create a custom management command, you can set up your app directory like this:

.. code-block:: text

    polls/
        __init__.py
        models.py
        management/
            __init__.py
            commands/
                __init__.py
                shell.py
        tests.py
        views.py

In this way, putting the app before the overridden shell app in :setting:`INSTALLED_APPS`, you will override the default shell command.

Overriding the Shell
====================

To override the command, you have to override the ``shell.Command.get_namespace`` in order to return a namespace that maps every name to the object imported. For example, you can edit ``polls/management/commands/shell.py`` to look like this::

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            from django.urls.base import resolve, reverse

            return {
                **super().get_namespace(),
                "resolve": resolve,
                "reverse": reverse,
            }

> This way, your overridden shell will import models from your apps and also the ``resolve`` and ``reverse`` methods so that you can use those without import.

Reviewer 2: ```suggestion
This way, your customized shell will import models from your apps along with the ``resolve`` and ``reverse`` methods, allowing you to use them without needing additional imports.
```

.. note::

>     If you want your shell not to import models, you can also omit calling the ``**super().get_namespace()`` method when returning the namespace because it can be empty.

Reviewer 2: ```suggestion
    If you prefer your shell without any auto-imported model, you can omit calling the ``**super().get_namespace()`` method when returning the namespace, as it can be empty.
```

Reviewer 2: not sure if this note is needed, but maybe a nice-to-have.
